bio/bio.go

```diff
@@ -0,0 +1,195 @@
+}
+
+/******************************************************************************
+
+Higher level parse
+
+******************************************************************************/
+
+type Parser[DataType fasta.Record, DataTypeHeader fasta.Header] struct {
+	LowLevelParser LowLevelParser[DataType, DataTypeHeader]
+}
+
+var emptyParser Parser[fasta.Record, fasta.Header] = Parser[fasta.Record, fasta.Header]{}
+
+func NewParser(format Format, r io.Reader) (*Parser[fasta.Record, fasta.Header], error) {
+	var maxLineLength int
+	switch format {
+	case Fasta:
+		maxLineLength = DefaultMaxLengths[format]
+	}
+	return NewParserWithMaxLine(format, r, maxLineLength)
+}
+
+func NewParserWithMaxLine(format Format, r io.Reader, maxLineLength int) (*Parser[fasta.Record, fasta.Header], error) {
+	switch format {
+	case Fasta:
+		return &Parser[fasta.Record, fasta.Header]{LowLevelParser: fasta.NewParser(r, maxLineLength)}, nil
+	}
+	return nil, nil
+}
+
+/******************************************************************************
+
+Parser read functions
+
+******************************************************************************/
+
+func ReadWithMaxLine(format Format, path string, maxLineLength int) (*Parser[fasta.Record, fasta.Header], error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return &emptyParser, err
+	}
+	return NewParserWithMaxLine(format, file, maxLineLength)
+}
+
+func Read(format Format, path string) (*Parser[fasta.Record, fasta.Header], error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return &emptyParser, err
+	}
+	return NewParser(format, file)
+}
+
+/******************************************************************************
+
+Parser higher-level functions
+
+******************************************************************************/
+
+func (p *Parser[DataType, DataTypeHeader]) Next() (DataType, int64, error) {
+	return p.LowLevelParser.Next()
+}
+
+func (p *Parser[DataType, DataTypeHeader]) Header() (DataTypeHeader, int64, error) {
+	return p.LowLevelParser.Header()
+}
+
+func (p *Parser[DataType, DataTypeHeader]) ParseN(countN int) ([]DataType, error) {
+	var records []DataType
+	for counter := 0; counter < countN; counter++ {
+		record, _, err := p.Next()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				err = nil // EOF not treated as parsing error.
+			}
+			return records, err
+		}
+		records = append(records, record)
+	}
+	return records, nil
+}
+
+func (p *Parser[DataType, DataTypeHeader]) Parse() ([]DataType, error) {
+	return p.ParseN(math.MaxInt)
+}
+
+func (p *Parser[DataType, DataTypeHeader]) ParseAll() ([]DataType, DataTypeHeader, error) {
+	header, _, err := p.Header()
+	if err != nil {
+		return []DataType{}, DataTypeHeader{}, err
+	}
+	data, err := p.Parse()
+	if err != nil {
+		return []DataType{}, DataTypeHeader{}, err
+	}
+	return data, header, nil
+}
+
+func (p *Parser[DataType, DataTypeHeader]) ParseConcurrent(channel chan<- DataType) error {
```

Comment:
As written, this non-current and blocking. You probably want to wrap the for loop in a `go func(){ ... }()` to make it nonblocking and call it `ParseToChannel()` instead. Make sure to note whether or not it is block in that case.

Reply:
Concurrency in term of parsing / IO is kind of a lie though, because you are literally reading the file with ` io.Reader`. I do agree though, bad name, I like the name `ParseToChannel`

(removed the `close(channel)` for below stuff)

The change to `go func(){}()` would make it nonblocking, but also make error checking more difficult. I'm thinking of using this with something like [errgroup](https://pkg.go.dev/golang.org/x/sync/errgroup), so that you can begin parsing multiple files while still having error handling in a pretty simple-to-understand way.

In the case that you are parsing one file, IO will likely be your limiting factor, and a blocking function is pretty easy to use and understand how it works. As you need to scale, you can just wrap the overall function in `go func(){}()`.


I am open to making it non-blocking tho - is there any particular reason you have in mind for making it non-blocking?